File: 2023/day03a.py
import re

total = 0

with open(0) as f:
    txt = f.read()
    lines = txt.splitlines()
    symbols = set(re.findall("(?!\d+|\.|\n).", txt))

    for row, line in enumerate(lines):
        for match in re.finditer("\d+", line):
            top = max(0, row - 1)
            left = max(0, match.start() - 1)
            bottom = min(len(lines), row + 1) + 1  # +1 because range() is exclusive
            right = min(len(line), match.end()) + 1  # +1 because range() is exclusive

            for schematic in lines[top:bottom]:
                if any(substr in schematic[left:right] for substr in symbols):
                    total += int(match.group())
                    break


print(total)

File: go/2023/day03a.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"unicode"
)

func main() {
	fmt.Println(solve())
}

func solve() int {
	var total int

	re := regexp.MustCompile("\\d+")

	lines := processStdIn()
	for row, line := range lines {
		for _, match := range re.FindAllStringIndex(line, -1) {
			//   -1012
			// -1 .*..
			//  0 .35.
			// +1 ....
			top := max(0, row-1)
			left := max(0, match[0]-1)
			bottom := min(len(lines)-1, row+1)
			right := min(len(line)-1, match[1])

			for _, schematic := range lines[top : bottom+1] {
				for _, char := range schematic[left : right+1] {
					if !(unicode.IsDigit(char) || string(char) == ".") {
						num, _ := strconv.Atoi(line[match[0]:match[1]])
						total += num
					}
				}
			}
		}
	}

	return total
}

func processStdIn() []string {
	var lines []string
	scanner := bufio.NewScanner(bufio.NewReader(os.Stdin))
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines
}
